Added web profiler and toolbar

Nyholm opened the pull request:
As mentioned in #3. 

This PR implements a web debug toolbar. It uses the HistoryPlugin.

dbu commented on Collector/MessageJournal.php:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function collect(Request $request, Response $response, \Exception $exception = null)
```
is this intentionally empty? if so, please add a comment why it is there and empty.

Nyholm replied:
Okey, thank you. 

A upcoming PR maybe want to log which Symfony Request that made which Httplug Request. But it should be empty for now.

dbu replied:
:+1:

dbu commented on DependencyInjection/HttplugExtension.php:
```diff
@@ -56,6 +61,10 @@ protected function configureClients(ContainerBuilder $container, array $config)
             }
 
+            if (isset($config['_inject_journal_plugin'])) {
+                array_unshift($arguments['plugins'], 'httplug.plugin.history');
```
what if somebody was already adding the history plugin? should we array_search ('httpplug.plugin.history') to avoid that?

dbu replied:
ah, read the doc now. maybe throw an InvalidConfigurationException if its explicitly specified? then again, why not allow users to activate the history plugin and do their own thing, even in prod mode if they feel its a good idea?

Nyholm replied:
We cant do a search/replace. People should not use "our" namespace when defining services so they would create a service like `my_history_plugin`. So it is impossible to do a search because we dont know what to search for... 

If someone uses our service, then the history will be logged twice. The readme says it is added automatically. We should not bother. 

What we should care about is that what if they are not happy with our formatter? But I think that can be solved by #17.

dbu replied:
ah, got it. maybe we should use a more specific service name for this. its not "the" history plugin, but a history plugin linked to the message journal.

Nyholm replied:
I'll make this service "internal" instead.. Im using service id `httplug.collector.history_plugin`

dbu commented on README.md:
```diff
@@ -51,14 +51,16 @@ For information how to write applications with the services provided by this bun
 | httplug.client.[name] | This is your Httpclient that you have configured. With the configuration below the name would be `acme_client`.
 | httplug.client | This is the first client configured or a client named `default`.
-| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services. 
+| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.history<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services.<br><br>*The History plugin is only availab ...
```
s/build in/built in/

dbu replied:
s/that lives/that live/  (plural)

dbu commented:
very cool! i am impressed how simple this looks!

Nyholm commented:
> very cool! i am impressed how simple this looks!

Thank you. So am I. I barley wrote any code just some clue to make `Httplug\*` classes talk with Symfony.

dbu commented on README.md:
```diff
@@ -51,14 +51,16 @@ For information how to write applications with the services provided by this bun
 | httplug.client.[name] | This is your Httpclient that you have configured. With the configuration below the name would be `acme_client`.
 | httplug.client | This is the first client configured or a client named `default`.
-| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services. 
+| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.history<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are built in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services.<br><br>*The History plugin is only availab ...
```
s/that lives/that live/

dbu commented on README.md:
```diff
@@ -51,14 +51,16 @@ For information how to write applications with the services provided by this bun
 | httplug.client | This is the first client configured or a client named `default`.
-| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are build in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services. 
+| httplug.plugin.content_length <br> httplug.plugin.decoder<br> httplug.plugin.error<br> httplug.plugin.history<br> httplug.plugin.logger<br> httplug.plugin.redirect<br> httplug.plugin.retry | These are built in plugins that lives in the `php-http/plugins` package. These servcies are not public and may only be used when configure HttpClients or services.<br><br>*The History plugin is only availab ...
 
```
lets make an explicit note on the history plugin. that plugin needs to be set up with a journal, so you need to define your own service for it.

Nyholm replied:
I've made it "internal". Configuration for plugins should be solved by #17

dbu commented on Resources/config/data-collector.xml:
```diff
@@ -0,0 +1,16 @@
+                 id="httplug"/>
+        </service>
+
+        <service id="httplug.plugin.history" class="Http\Client\Plugin\HistoryPlugin" public="false">
```
lets use a more specific service name here to prevent confusion with generally useable plugins.

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -72,6 +72,7 @@ public function getConfigTreeBuilder()
                         ->scalarNode('stream_factory')->defaultNull()->end()
                     ->end()
                 ->end()
+                ->booleanNode('toolbar')->defaultTrue()->end()
```
i would make toolbar an arrayNode with canBeEnabled and a message_formatter option to use a custom formatter. its not a plugin configuration for me. the fact that the history plugin is used is an implementation detail. if i configure my own services with history plugin instances, i need my own configuration for formatters anyways.

the other plugins can be configured and then used as they are in any client the user wants to use them.

Nyholm replied:
Agreed. But I would like `canBeDisabled` instead... It should be enabled by default.

dbu replied:
oh, you are right, yes. maybe add a description() saying that the toolbar is only active if the debug toolbar itself is active.

Nyholm replied:
Thank you. I've updated the PR

dbu commented on DependencyInjection/HttplugExtension.php:
```diff
@@ -27,6 +27,13 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('services.xml');
         $loader->load('plugins.xml');
         $loader->load('discovery.xml');
+        if ($config['toolbar']['enabled'] && $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
```
is this state of the art to detect whether the debug toolbar is active? what if somebody has it active in prod env or deactivated in dev env? i guess in the latter case, we just waste some cpu cycles in dev mode which is no big deal. the other might be a problem. or is debug toolbar hard coupled to kernel.debug?

Nyholm replied:
You are correct. I should redo this. 
What I really want to know is if the service `profiler` exits and if its configuration for `toolbar` is true. 

I'll investigate some more.

dbu replied:
maybe look at what other bundles that plug into the toolbar do

Nyholm replied:
> or is debug toolbar hard coupled to kernel.debug?

No it is not. To get the toolbar you need `framework.profiler` enabled and of course the WebProfilerBundle. There is nothing that ties it to kernel.debug. So this is technically wrong, but it works for most cases. Others are doing the same.

https://github.com/egeloen/IvoryHttpAdapterBundle/blob/master/DependencyInjection/IvoryHttpAdapterExtension.php#L62

I could use a compiler pass and check for the `web_profiler.debug_toolbar.mode` parameter if it is enabled or disabled. But I think that will complicate how I inject the HistoryPlugin because that would also be done in the compiler pass. 

Question: Are we happy with postponing support for the use case where you want the debug Httplug request via the toolbar with `kernel.debug: false`?

Nyholm replied:
> maybe look at what other bundles that plug into the toolbar do

I've looked at a few. It is common that you inject a logger or a collector to a decorating class. That injection is done in a compiler pass. Some bundles (like [GeocoderBundle](https://github.com/geocoder-php/BazingaGeocoderBundle) do always inject the logger, no matter debug or environment.

sagikazarmark replied:
It's always a good idea to inject logger. Fingers Crossed handler can
decide then when and what to log.

Tobias Nyholm <notifications@github.com> ezt írta (2016. január 8., péntek):

> In DependencyInjection/HttplugExtension.php
> <https://github.com/php-http/HttplugBundle/pull/19#discussion_r49206816>:
>
> > @@ -27,6 +27,13 @@ public function load(array $configs, ContainerBuilder $container)
> >          $loader->load('services.xml');
> >          $loader->load('plugins.xml');
> >          $loader->load('discovery.xml');
> > +        if ($config['toolbar']['enabled'] && $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
>
> maybe look at what other bundles that plug into the toolbar do
>
> I've looked at a few. It is common that you inject a logger or a collector
> to a decorating class. That injection is done in a compiler pass. Some
> bundles (like GeocoderBundle
> <https://github.com/geocoder-php/BazingaGeocoderBundle> do always inject
> the logger, no matter debug or environment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/php-http/HttplugBundle/pull/19/files#r49206816>.
>

Nyholm replied:
> It's always a good idea to inject logger. Fingers Crossed handler candecide then when and what to log.

Yes, I meant more of a DataCollector to collect data for the profiler. That is injected all the time

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -72,8 +71,14 @@ public function getConfigTreeBuilder()
+                    ->canBeDisabled()
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->scalarNode('formatter')->defaultValue('httplug.collector.message_journal')->end()
```
erm, i think this default value is wrong. i think it should defaultNull and if set it must be a service of a class that implements Http\Message\Formatter - or am i confused?

Nyholm replied:
No, sorry. Im confused.. You are correct.

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -34,8 +34,7 @@ public function getConfigTreeBuilder()
                         || !empty($v['classes']['uri_factory'])
-                        || !empty($v['classes']['stream_factory'])
-                    ;
+                        || !empty($v['classes']['stream_factory']);
```
is this styleci? my idea was that adding another check only does a diff on one line, like trailing comma in arrays.

Nyholm replied:
Yeah, this is styleci

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
+                        ->scalarNode('formatter')->defaultNull()->end()
+                    ->end()
+                ->end()
+            ->end();
```
same logic here, if we add another section, there will be an unnecessary diff on this line.

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
-            ->end()
-        ;
+                ->arrayNode('toolbar')
+                    ->canBeDisabled()
```
regarding kernel.debug: could we make enabled a simple field and default it to the value kernel.debug? then people could just explicitly enable/disable it if they explicitly enabled the toolbar despite kernel.debug being false. if kernel.debug is not available here, we could have a choice of `true`, `false` and `"auto"`, with auto meaning the DI extension classs looks at kernel.debug.

Nyholm replied:
Sounds like a plan

dbu commented on DependencyInjection/HttplugExtension.php:
```diff
@@ -27,6 +27,17 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('plugins.xml');
         $loader->load('discovery.xml');
+
+        if ($config['toolbar']['enabled'] && $s = $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
```
so here you would do something like:

$enabled = is_boolean($config['toolbar']['enabled']) ? $config['toolbar']['enabled'] : $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug');
if ($enabled) {
...

Nyholm replied:
Thanks! I've updated the PR.

sagikazarmark commented on README.md:
````diff
@@ -51,15 +51,16 @@ For information how to write applications with the services provided by this bun
 ```yaml
 httplug:
-  clients:
+  clients: 
````
Is this a trailing space?

sagikazarmark commented on Resources/views/webprofiler.html.twig:
```diff
@@ -0,0 +1,75 @@
+{% endblock %}
+
+{% block panel %}
+    <h2>Httplug</h2>
```
HTTPlug :smile:

sagikazarmark commented on Tests/Resources/Fixtures/config/full.php:
```diff
@@ -1,16 +1,16 @@
+        'stream_factory'  => 'my_stream_factory',
+    ],
+    'classes' => [
+        'client'          => 'Http\Adapter\Guzzle6\Client',
```
Is this styleci as well? AFAIK symfony preset unaligns double arrows, doesn't it?

dbu replied:
Tests is ignored by styleci i think. not sure if we really want to ignore them. consistent CS in the tests sounds like a valuable thing to have.

sagikazarmark replied:
Ah, I see. Well, the reason why I ignored tests by default was that spec tests are snake cased and function names does not contain visibility. I agree it would make sense here, so let's just remove the test paths from exclusions.

Nyholm replied:
Yes, That is styleci. 
Ignoring sounds like a good idea.

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
-        ;
+                ->arrayNode('toolbar')
+                    ->addDefaultsIfNotSet()
+                    ->info('Extend the debug profiler with inforation about requests. This could only be active when the kernel is in debug mode.')
```
lets remove the second phrase

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -72,8 +71,15 @@ public function getConfigTreeBuilder()
+                    ->addDefaultsIfNotSet()
+                    ->info('Extend the debug profiler with inforation about requests. This could only be active when the kernel is in debug mode.')
+                    ->children()
+                        ->enumNode('enabled')->values([true, false, 'auto'])->defaultValue('auto')->end()
```
and add an info() here saying that by default, the toolbar is active when kernel.debug is true, but you can force it to on or off with this option.

Nyholm commented:
Thank you for your feedback. I've updated the PR.

sagikazarmark commented:
Sorry, CS fixes probably introduced a few conflicts.

Nyholm commented:
No worries, I've updated the PR.

sagikazarmark commented:
Cool. How far is it? I am eager to try it in a project. :smile:

Nyholm commented:
I'm done with the implementation, I've received a lot feedback and made some improvements. I have addressed all the issues that were brought up. Now Im waiting for @dbu to give this PR the final thumbs up.

sagikazarmark commented:
Cool.

dbu commented:
thanks a lot for this one, good job!

sagikazarmark commented:
:+1:

Nyholm commented:
Thank you for merging. I'm also very exited to try this in a project =)

dbu commented on DependencyInjection/HttplugExtension.php:
```diff
@@ -54,6 +66,10 @@ protected function configureClients(ContainerBuilder $container, array $config)
             }
 
+            if (isset($config['_inject_collector_plugin'])) {
+                array_unshift($arguments['plugins'], 'httplug.collector.history_plugin');
```
actually, adds the plugin first. does that mean we log before anything else? that will mean that if we have the redirect plugin active, we will not see that a redirect happened. then again, if redirect was activated, we might not care about it.
but it also means we will not log the request exactly as it was sent, with base url or credentials or whatever was added by the plugins. do we maybe need a flag to decide whether to add to the beginning or to the end?

Nyholm replied:
If you have the CachePlugin and put the HistoryPlugin last you will not see anything... But that might be okey, since no request was actually made.

dbu replied:
true. hm. it would be cool to see whether requests where served from
cache or the actual backend. we could try to have two history plugins
and see whether a request goes to the client or not, and also see
whether there is a redirect to serve one answer.

dbu replied:
created #26 for this idea. i guess what we have now is ok, but the other thing sure would be cool.
